### ISSUE TITLE
What traits/aspects of the cancer data led you to choose the lenses? 

### ISSUE THREAD

**Issue by karinsasaki** [scikit-tda/kepler-mapper#134]
I am looking at the Cancer python notebook and, although I understand how the lenses project the data, I am still wondering how the two lenses (isolation forest and l2norm) were chosen to analyse this specific data?

More precisely, I want to know if there were some specific aspects of the data that made you think those two lenses were appropriate (e.g. do these lenses have some biological significance? or do you, a priori, know something about the "shape" of the data that tells you theses lenses are appropriate?); or did you do a sort of "grid search" of the choice of lenses and parameters for the map.mapper function (specifically the nr_cubes, overlap_perc, and the clusterer and its parameters)? If you did do a grid search, did you choose the "best performing" combination qualitatively (by looking at the topological graphs generated) or did you use some kind of quantitative measure (and what is it)?

Thank you so much for any guidance!

**Comment by carlomotta** [comment]
This is a great question, I would really like to read an answer!

**Comment by karinsasaki** [comment]
Looking at the different effects different lenses have on the shape of the final topological graph, I have concluded that a reasoning behind a choice of a combination of lenses for a particular dataset is:

1. Lenses that make biological sense; in other words, lenses that highlight special features in my data that I know about. I imagine that in the case of the cancer data, using an **anomaly score** (in this case calculated using the IsolationForest from sklearn) makes biological sense.
2. Lenses that somehow disperse the data, as opposed to clustering many points together (see below).

-----------------

Let's look at the cancer data:

This is the image of lens1 (the anomaly score using the IsolationForest):
![Image of lens1](https://user-images.githubusercontent.com/18656780/53474558-b2f8b200-3a6d-11e9-895e-1e6cfc1b0658.png)

(obtained using:
```
plt.scatter(lens1,np.zeros(len(lens1)))
plt.title('Image of lens1')
```
)

This is the image of lens2 (the l2norm):
![Image of lens2](https://user-images.githubusercontent.com/18656780/53474576-c146ce00-3a6d-11e9-9967-0748ddbd31e7.png)

This is the scatterplots for the two lenses color coded by the diagnosis:
![lens1 vs lens2](https://user-images.githubusercontent.com/18656780/53479012-6ebedf00-3a78-11e9-82f8-acb977008892.png)

Using PCA I can also create another lens as follows:
```
from sklearn.decomposition import PCA
lens3 = mapper.fit_transform(X, projection=PCA(n_components=3), scaler=None)
```

This is the scatterplot of the first two PCA eigenvectors against each other:
![lens3[:,0] vs lens3[:,1]](https://user-images.githubusercontent.com/18656780/53479061-8b5b1700-3a78-11e9-8e78-6fb19f8362ec.png)

and this is the scatter plot between lens1 (anomaly score using IsolationForest) and lens3[:,0] (the first Eigenvector of the PCA):
![lens1 vs lens3[:,0]](https://user-images.githubusercontent.com/18656780/53479249-e1c85580-3a78-11e9-95d3-f240ed4b9a9f.png)

Notice it looks so similar (or the same(?)) as lens1 vs lens2.

And if we generate the topological graphs, this is what we get with the different lenses:

_lens = np.c_[lens1, lens2]_
![lens1, lens2](https://user-images.githubusercontent.com/18656780/53479502-7f238980-3a79-11e9-804e-14d73d673496.png)

_lens = np.c_[lens1, lens3[:,0]]_
![lens1, lens3](https://user-images.githubusercontent.com/18656780/53479522-864a9780-3a79-11e9-945e-6bb37d6f109b.png)

_lens = np.c_[lens3[:,0],lens3[:,1]]_
![PCS lenses](https://user-images.githubusercontent.com/18656780/53479506-80ed4d00-3a79-11e9-9fd3-772bef319653.png)

Again, using _lens = np.c_[lens1, lens2]_ and _lens = np.c_[lens1, lens3[:,0]]_ generates (what visually seems to be the same) topological graph. [As a side note, it would be nice to be able to compare topological graphs somehow quantitatively and not just visually, e.g. by calculating homology or some other way - any thoughts on this?]

This is what we get if we generate the topological graph using only lens 1:
![lens1](https://user-images.githubusercontent.com/18656780/53479673-d1fd4100-3a79-11e9-9844-84dfe6f9cf90.png)

So for me, the conclusion for the cancer dataset is that using the anomaly score gives the global shape of the data and using another lens in addition, such as the l2norm or the first PCA eigenvector, gives a less granular view of the different regions of the topological graph.

(Note, the rest of the parameter values are the same for all topological graphs (e.g. number of hypercubes, clustered, percentage overlap); I have only changed the lens used.)

**Comment by sauln** [comment]
This a great write-up! What do you think about incorporating it into the documentation as a tutorial page?

**Comment by karinsasaki** [comment]
@sauln Yes, I'd be happy to! I'll have a look at the other tutorials to get a better idea on structure. 

Can you tell me if I need to make a PR or send it over in some other way?

**Comment by carlomotta** [comment]
@karinsasaki thank you for your nice explanation.

> * What does the graph tell me about the data?
Something that may be useful: if you color the nodes based on your target value y, you may notice hotspot regions that are characterized by a larger presence of positively targeted observations. For instance, look at the right flare on the below picture.

<img width="791" alt="screen shot 2019-03-01 at 17 57 20" src="https://user-images.githubusercontent.com/22749071/53653445-9a45f300-3c4b-11e9-8e66-6598ad3f9159.png">

**Comment by sauln** [comment]
Most of the other tutorials were implemented in Jupiter notebooks and then converted to docs pages during the build process. That's fine if you'd like to go with that. Also, writing in RST is also fine.

a PR would preferable. Once you submit it, I can do a code review/copy edit pass.

**Comment by karinsasaki** [comment]
@sauln sorry it took me a while! I needed to finish other stuff first. I've sent over a PR - let me know if you think of any changes or improvements!

@carlomotta yea, I also though about colouring by the labels (y)! I suppose the quantifying the points on that flare would make sense.

**Comment by carlomotta** [comment]
@karinsasaki I believe that it would be highly beneficial an automated process to:

- _identifying_ flares and hotspots (i.e. region characterized by a high level of a certain variable)
- _quantifying_ and _categorize_ the relative nodes/records
- _qualifying_ those records

**Comment by karinsasaki** [comment]
@carlomotta do you have a paper (or something else) that showcases these three things being done clearly for a dataset (whatever the dataset is, it doesn't matter to me). I would love to see a clear example of, not only how to use the mapper algorithm, but also how to make clear conclusions from the topological graph, that are relevant for the data.

**Comment by carlomotta** [comment]
@karinsasaki please, check out [this paper](https://s3.amazonaws.com/cdn.ayasdi.com/wp-content/uploads/2016/01/13110816/UCSF-Nature-6.pdf).

By _qualifying_ a group in a topological network, I mean assessing which features in the original multidimensional space are more "responsible" of the data belonging to that group. For instance, you can perform statistical tests (e.g. Kolmogorov–Smirnov test) to compare the distribution of each feature within a group and on all the complementary nodes.

**Comment by carlomotta** [comment]
@karinsasaki please, also have a look at [this paper](https://www.nature.com/articles/srep01236?__hstc=13887208.d8433f44817d66028732a6792cd78bb4.1471910400048.1471910400050.1471910400051.2&__hssc=13887208.1.1471910400051&__hsfp=1773666937).

**Comment by MLWave** [comment]
I am the author of that example, and not a domain expert on cancer analysis. There must be better lenses and approaches available, perhaps in: [0](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5771419/) [1](https://www.nature.com/articles/ng.3806) 

This is the original write-up: http://mlwave.github.io/tda/breast-cancer-writeup.html

I treated the problem as unsupervised anomaly detection. Isolation Forest is one of the best anomaly detectors I know of, and for L2-norm I reasoned that it would give anomalous points a different norm than typical points. 2-dimensional lenses are a bit richer.

I think the original Lum et al. used L2-norm over a distance matrix, not over the raw vectors. I'd probably try that now (to use `distance_matrix` and `l2norm` or `knn_distance_5`). Together with `sklearn.cluster.agglomerativeClustering`.

> Kolmogorov–Smirnov test

You could also use the label distribution, whitebox classifier accuracy, or entropy of the clusters/nodes to get a fitness score for the graph. Then you could automatically pick from all anomaly detectors available. Though individually, lenses may give a different view all together. 

For demonstrations like this, I am looking for richly connected graphs with good division.

**Comment by MLWave** [comment]
The original approach from Lum et al. is:

lens1: The target variable (survival or relapse)
lens2: L-infinity centrality

Lens1 then causes the layout to separate on the target. Lens2 gives a centrality measure. You can then find the different types of survival or relapse, and flares should form more clearly (each signifying a different type of departure from the centrality).

Instead of the target variable, you may also use the predictive error of an ML model. Then you can study different types of failure, a la "Fibres of Failure".

**Comment by karinsasaki** [comment]
@carlomotta Thank you for the links! I'll check them out.

@MLWave Thanks so much for expanding on this example! It's great to know the thinking process.
